snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+		final BranchContext context, 
+		final List<ExtendedLocale> locales, 
+		final Collection<QueryExpression> inclusions, 
+		final Collection<QueryExpression> exclusions
+	) {
+		return conceptSetEvaluator.evaluateConceptSet(context, locales, resourceUri, inclusions, exclusions, PAGE_SIZE);
+	}
+
+	private Stream<String> evaluateEcl(final BranchContext context, final String eclExpression) {
+		return evaluator.evaluateEcl(context, eclExpression, PAGE_SIZE);
+	}
+
+	private Set<String> evaluateEclToSet(final BranchContext context, final String eclExpression) {
+		return evaluateEcl(context, eclExpression).collect(Collectors.toSet());
+	}
+
+	private List<QueryExpression> optimizePinned(final BranchContext context, final Collection<QueryExpression> expressions) {
+
+		// Find unique pinned queries
+		final Map<Set<String>, String> pinnedQueries = expressions.stream()
+			.filter(clause -> clause.isPinned())
+			.collect(Collectors.toMap(
+				clause -> evaluateEclToSet(context, clause.getQuery()), 
+				clause -> clause.getQuery(),
+				BinaryOperator.minBy(Comparator.comparing(String::length)))); // shorter expression wins
+
+		if (pinnedQueries.isEmpty()) {
+			return List.of();
+		}
+
+		return pinnedQueries.values()
+			.stream()
+			.map(query -> new QueryExpression(IDs.base62UUID(), query, true))
+			.collect(Collectors.toList());
+	}
+
+	private void applyInclusions(final BranchContext context, final List<QueryExpression> newInclusions) {
+		applyInclusions(context, newInclusions, true);
+	}
+
+	private void applyInclusions(final BranchContext context, final List<QueryExpression> newInclusions, final boolean processExclusions) {
+
+		if (newInclusions.isEmpty()) {
+			return;
+		}
+
+		// Add inclusions to our optimized list
+		optimizedInclusions.addAll(newInclusions);
+
+		// Evaluate queries; remove concepts that are members of the original set from "conceptsToInclude"; track extras in "conceptsToExclude"
+		newInclusions.stream()
+		.map(clause -> evaluateEclToSet(context, clause.getQuery()))
+		.forEachOrdered(includedIds -> {
+			final Set<String> expected = Sets.intersection(includedIds, conceptSet);
+			conceptsToInclude.removeAll(expected);
+
+			if (processExclusions) {
+				final Set<String> unexpected = Sets.difference(includedIds, conceptSet);
+				conceptsToExclude.addAll(unexpected);
 			}
-			
-			// optimization is a "net win" if we can remove at least two non-pinned clauses from the original (pinned clauses cannot be optimized)
-			final List<QueryExpression> nonPinnedMembersForParent = List.copyOf(membersByAncestor.get(parentId)
-					.stream()
-					.filter(ex -> !ex.isPinned())
-					.collect(Collectors.toList()));
-			
-			if (nonPinnedMembersForParent.size() <= 1) {
-				continue;
+		});
+
+		newInclusions.clear();
+	}
+
+	private void filterRefinementsForInclusion(final SnomedRelationshipStats relationshipStats) {
+		// a) Less than 95% precision
+		relationshipStats.filterByPrecision(0.95f);
```

Comment:
Would it make sense to move any of these magic numbers to constants? Or have them configurable?

Reply:
We can expose any of the parameters as `Option` keys, and pick up defaults from eg. `SomedCoreConfiguration`, but I'm not sure who would want to adjust them. An "advanced optimize" dialog would look like a studio mixing board.

Reply:
It's okay to move them to named constants. I'm unsure whether having them configurable externally would be great or not.